### DISCUSSION TITLE
notation rc.2 scope

### DISCUSSION THREAD

**Discussion by yizha1** [notaryproject/notaryproject#206] [General]
Here are the candidates for notation rc.2 from user scenarios point of view.

- [ ] Inspect signatures
- https://github.com/notaryproject/notation/issues/172
- https://github.com/notaryproject/notaryproject/issues/95

- [ ] Support OCI Image manifest to store signatures
- https://github.com/notaryproject/notation/issues/444

- [ ] Use CLI to manage Trust Policy 
- https://github.com/notaryproject/roadmap/issues/39

- [ ] Signature filtering
- https://github.com/notaryproject/notation-go/issues/197
- https://github.com/notaryproject/notation-go/issues/70

- [ ] Offline Sign
- https://github.com/notaryproject/roadmap/issues/15

- [ ] Certificate revocation
- https://github.com/notaryproject/roadmap/issues/11

- [ ] Signature revocation

- [ ] Signature RFC3161 timestamping

- [ ] Tag signing

- [ ] Quality: E2E Testing automation
- https://github.com/notaryproject/notation-go/issues/52
- https://github.com/notaryproject/roadmap/issues/47

**Comment by yizha1** [comment]
Copied from @iamsamirzon 's [slack message](https://cloud-native.slack.com/archives/CQUH8U287/p1668809865951419?thread_ts=1668709044.207749&cid=CQUH8U287)

Fall back support - 1
Notation Inspect signatures - 2
Complete Debug option  -2
Signature filtering - 2
E2E test cases - 2
CLI cmds for TP Management - 2
Timeout Implementation - 3
API Documentation - 4
CLI Documentation - 4
Pull and Push command feature in notation - unknown (edited)

**Comment by yizha1** [comment]
**Summary**

Notation rc.2 scope consists of
- rc.1 leftovers
- New features
- rc.1 bugs reported after the release (not clear for now, but need reserve time for it)

**Backlogs**

| Title | Github issue | Priority | Comments | cost estimation (Impl+test+PR review) |
| -------- | ------ | -------- | -------- | ------- |
| Fall back support |  [notation#444](https://github.com/notaryproject/notation/issues/444)   |  P0    |  rc.1 leftover, but new scenario | 2w 
| noation offline sign/verify | [roadmap#15](https://github.com/notaryproject/roadmap/issues/15) | P0 | New feature | 4w
| E2E testing | [roadmap#47](https://github.com/notaryproject/roadmap/issues/47), [roadmap#63](https://github.com/notaryproject/roadmap/issues/63) | P0 | rc.1 leftover, Quality improvement | 3w
| CLI cmds for TP management | [roadmap#39](https://github.com/notaryproject/roadmap/issues/39) | P0 |  New feature | 2w 
| inspect signatures |[notation#172](https://github.com/notaryproject/notation/issues/172) | P1 |  New feature  | 1w
| debug option | [notation#300](https://github.com/notaryproject/notation/issues/300) | P1 -> P0 | rc.1 leftover, troubleshooting, -> sign and verify debug logs should be implemented in rc.1| 1w
| signature filtering | [notation-go#197](https://github.com/notaryproject/notation-go/issues/197) | P1 | New feature |  2w
| certificate revocation | [roadmap#11](https://github.com/notaryproject/roadmap/issues/11), [roadmap#60](https://github.com/notaryproject/roadmap/issues/60) | P2 | New feature | 8w ~ 12w |
| Timestamp signature authority | [roadmap#59](https://github.com/notaryproject/roadmap/issues/59) | P2 | New feature | 8w ~ 12w
| Timeout implementation | [notation-go#201](https://github.com/notaryproject/notation-go/issues/201) | P2 | rc.1 leftover |  
| API documentation | [notation-go#189](https://github.com/notaryproject/notation-go/issues/189) | P3 | New feature| 2w ~ 4w
| CLI manual documentation | [website#88](https://github.com/notaryproject/notaryproject.dev/issues/88) | P3 | Documentation improvement | 2w ~ 4w
| pull and push command | [notation#245](https://github.com/notaryproject/notation/issues/245), [notation#246](https://github.com/notaryproject/notation/issues/246) | P4 | any user scenarios we want to enable? could be offline sign/verify| 2w
| Tag signing | [roadmap#16](https://github.com/notaryproject/roadmap/issues/16) | TBD | new feature
| code clean-up| [notation#392](https://github.com/notaryproject/notation/issues/392), [notation#446](https://github.com/notaryproject/notation/issues/446), [notation-go#80](https://github.com/notaryproject/notation-go/issues/80), [notation-core-go#96](https://github.com/notaryproject/notation-core-go/issues/96)| TBD | rc.1 leftover, suggest P0 | 1w
|  Behavior of VerificationOutcome | [notation-go#214](https://github.com/notaryproject/notation-go/issues/212)| TBD | rc.1 leftover | 2w ~ 3w
| VerifyOptions.MaxSignatureAttempts | [notation-go#212](https://github.com/notaryproject/notation-go/issues/212) |TBD | rc.1 leftover |
| Release Automation via GH actions | [roadmap#66](https://github.com/notaryproject/roadmap/issues/66) | TBD | rc.1 leftover | 
| Threat Modeling | [notation#391](https://github.com/notaryproject/notation/issues/391), [notation#409](https://github.com/notaryproject/notation/issues/409) |TBD | rc.1 leftover | 2w ~ 4w
| notation list pagination | [notation#403](https://github.com/notaryproject/notation/issues/403) | TBD | rc.1 leftover|
| Support expiry duration in days | [notation#423](https://github.com/notaryproject/notation/issues/423) | TBD| rc.1 leftover| 1w
| Update plugin spec to support extended attributes| [notation#201] (https://github.com/notaryproject/notaryproject/issues/201)| TBD | rc.1 leftover|


**TODOs**
- Align priorities including the format of priority using `P0` or `1`. (**Done**)
- Cleanup/create/update issues in `roadmap repo` based scope, so that we have a central place to manage the roadmap issues, otherwise the issues will be in different repos (notation, notation-go, notation-core-go). Align what kinds of issues are roadmap issues (User scenarios?) --> then we can create a roadmap board from roadmap repo issues to show what has been done, what is planned, what is in the backlogs, so that user can easily understand `notation project`
- Need input on the cost estimation, then we can know what is reasonable scope for rc.2 and post-rc.2 releases.
- code freeze check-in before release.
- Documentations or blogs for rc.1, since it is a major milestone. (**Feynman is on it**)
- Decision on release cadence: one month or two months? One month seems a bit tight considering PR review lead time.

/cc: @shizhMSFT  @FeynmanZhou @toddysm @iamsamirzon @vaninrao10 @priteshbandi

**Reply by FeynmanZhou** [reply]
Can you add two links to API and CLI documentation to this table? 

- API documentation: https://github.com/notaryproject/notation-go/issues/189  
- CLI documentation https://github.com/notaryproject/notaryproject.dev/issues/88

**Reply by FeynmanZhou** [reply]
For notation pull & push commands, they are deprecated in the alpha release. These two scenarios can also be covered by ORAS CLI. Is it still valid to add them back to RC.2?

**Comment by yizha1** [comment]
### rc.2 release
**Date**: Jan 6, 2023
**Scope**:
- Fall back support, 2 weeks + spec ready first, P0
- E2E testing, 3 weeks, P0 ( phased approach)
- debug option, <1 week for leftovers
- code clean-up, 1 week

**Reply by vaninrao10** [reply]
**Date Proposed for rc.2 release** - Jan 16, 2023

Couple of more items added for the scope of rc.2
- Semantic versioning, P0 (1 week)
- Inspect Command, P1, (3-4 weeks)
- Attestation [https://github.com/notaryproject/notaryproject/issues/201]. P1 (2 weeks)
- Revocation Spec - Only Spec (P2), (1 week)
- Review PRs

**Comment by yizha1** [comment]
rc.2 scope based on the discussion with @vaninrao10 :

**Target date** 

Jan 6, 2023

**Scope**

| Title | GitHub issue | Priority | need spec? | cost estimation (implement + test) | PR review | Total | comment
| -------- | ------ | -------- | -------- | -------| --------| -------| -------- |
| *Fall back support | https://github.com/notaryproject/notation/issues/444 | P0 |  Yes | 2w | 2w | 4w |  | 
| *E2E testing |https://github.com/notaryproject/roadmap/issues/47, https://github.com/notaryproject/roadmap/issues/63 | P0 | No | 1.5w | 1.5w | 3w | Phased approach, may not be complete in rc.2 | 
| Apply semantic versioning for plugins | https://github.com/notaryproject/notation-go/issues/102 | P0 | Yes  |1w |  | | need spec first
| CLI cmd for TP Management (CLI spec) | https://github.com/notaryproject/roadmap/issues/39 | P0 | Yes | 1w | 1w | 2w |only CLI spec in rc.2 
| Inspect signatures | https://github.com/notaryproject/roadmap/issues/74 | P1 | Yes | 1w | | | 
| *debug option | https://github.com/notaryproject/roadmap/issues/71 | P0| No | 3d | 3d| 6d | rc.1 leftovers |
| Revocation (spec) | https://github.com/notaryproject/roadmap/issues/11  | P2 |  Yes | 4d | | | only spec in rc.2 | 
| Attestation | https://github.com/notaryproject/notaryproject/issues/201, https://github.com/notaryproject/roadmap/issues/67 | P1 |
| Sign/verify offline artifacts (spec) | https://github.com/notaryproject/roadmap/issues/15 | P0 | Yes| | | | start spec work in rc.2 |
